File: front/src/socialComponentMonitoring/model/SocialComponentMonitoring.js
import {comments_api_adapter, createComments} from "comment/model";

class SocialComponentMonitorings extends Array {}

const social_component_monitoring_api_adapter = socialComponentMonitoring => {
    if (socialComponentMonitoring["comments"]) {
        socialComponentMonitoring["comments"] = createComments(
            comments_api_adapter(socialComponentMonitoring["comments"])
        );
    }

    return socialComponentMonitoring;
};

const social_component_monitoring_view_adapter = socialComponentMonitoring => {
    delete socialComponentMonitoring["folder"];
    delete socialComponentMonitoring["created_by"];
    delete socialComponentMonitoring["created_at"];
    delete socialComponentMonitoring["updated_at"];
    delete socialComponentMonitoring["updated_by"];

    delete socialComponentMonitoring["comments"];

    return socialComponentMonitoring;
};

const social_component_monitorings_api_adapter = socialComponentMonitorings => {
    return socialComponentMonitorings.map(social_component_monitoring_api_adapter);
};

const createSocialComponentMonitorings = (data = []) => {
    const socialComponentMonitorings = SocialComponentMonitorings.from(
        data,
        socialComponentMonitoring =>
            createSocialComponentMonitoring(socialComponentMonitoring)
    );
    return socialComponentMonitorings;
};

const createSocialComponentMonitoring = ({
    id = null,
    code = "",
    name = "",
    execution_status = "",
    execution_status_label = "",
    quality_status = "",
    quality_status_label = "",
    expected_end_date = "",
    real_end_date = "",
    progress_percentage = "",
    comments = [],
    project = null,
    folder = "",
    created_by = "",
    created_at = null,
    updated_by = "",
    updated_at = null,
} = {}) => {
    const publicApi = {
        id,
        code,
        name,
        execution_status,
        execution_status_label,
        quality_status,
        quality_status_label,
        expected_end_date,
        real_end_date,
        progress_percentage,
        comments,
        project,
        folder,
        created_by,
        created_at,
        updated_at,
        updated_by,
    };

    return Object.freeze(publicApi);
};

export {
    createSocialComponentMonitoring as default,
    createSocialComponentMonitorings,
    social_component_monitoring_api_adapter,
    social_component_monitorings_api_adapter,
    social_component_monitoring_view_adapter,
};

File: back/app/models/social_component_monitoring.py
from django.db import models

from app.base.models.base_models import ActiveManager, BaseEntityModelMixin
from app.models.comment import Comment
from documents.base.base_models import BaseDocumentModel


class SocialComponentMonitoring(BaseDocumentModel, BaseEntityModelMixin):
    class Meta(object):
        db_table = "social_component_monitoring"
        verbose_name = "Seguimiento de componente social"
        verbose_name_plural = "Seguimientos de componente social"

    objects = ActiveManager()

    code = models.CharField("Código", max_length=50)
    name = models.CharField("Nombre", max_length=255)
    execution_status = models.CharField("Estado de ejecución", max_length=20, null=True)
    quality_status = models.CharField("Estado cualitativo", max_length=20, null=True)

    expected_end_date = models.DateField("Fecha de finalización prevista", null=True)
    real_end_date = models.DateField("Fecha de finalización real", null=True)
    progress_percentage = models.DecimalField(
        "Porcentaje de progreso", max_digits=5, decimal_places=2, null=True
    )

    project = models.ForeignKey(
        "Project", on_delete=models.CASCADE, related_name="project_social_monitorings"
    )
    comments = models.ManyToManyField(Comment)

    @property
    def trainings(self):
        return self.social_component_monitoring_trainings.all()


def create_project_social_components(project, components):
    for component_code, component_config in components.items():
        component = SocialComponentMonitoring(
            code=component_code,
            name=component_config.get("name"),
            project=project,
            created_by=project.creation_user,
            updated_by=project.creation_user,
        )
        component.save()
